File: python/day1.py
from python.advent_utils import read_input


def keep_numbers(line: str) -> str:
    return "".join([c for c in line if c.isdigit()])


def part_1() -> str:
    inp = read_input("inputs/day01.txt")
    only_numbers = [keep_numbers(line) for line in inp]
    total = sum(int(n[0] + n[-1]) for n in only_numbers)
    print(total)
    return str(total)


if __name__ == "__main__":
    part_1()

File: zig/src/day1.zig
const print = @import("std").debug.print;
const utils = @import("advent_utils.zig");
const std = @import("std");

pub fn compose_number(input: []const u8) !usize {
    var first: ?usize = null;
    var second: []const u8 = undefined;

    const last = for (input) |c| {
        if (!std.ascii.isDigit(c)) {
            continue;
        }
        if (first == null) {
            first = try std.fmt.parseUnsigned(usize, &[1]u8{c}, 10);
        }
        second = &[1]u8{c};
    } else try std.fmt.parseUnsigned(usize, second, 10);

    return first.? * 10 + last;
}

pub fn part1(path: []const u8) !usize {
    var file_content = try utils.get_input(path, null);
    var input = std.mem.splitScalar(u8, file_content, '\n');
    var total: usize = 0;

    return while (input.next()) |line| {
        total += try compose_number(line);
    } else total;
}
